xla/service/gpu/matmul_utils.cc

```diff
@@ -484,48 +485,64 @@ StatusOr<bool> CanFoldTransposeOperandIntoDot(const HloInstruction& dot,
       grad_y);
 }
 
-namespace {
+auto GemmConfig::MatrixDescriptors(se::DeviceMemoryBase lhs_buf, 
+        se::DeviceMemoryBase rhs_buf, se::DeviceMemoryBase out_buf) const 
+                                                        -> DescriptorsTuple
+{
+  auto create = [](const se::gpu::MatrixLayout& m, se::DeviceMemoryBase data) {
+      // BLAS is column-major by default.
+    auto trans = (m.order == se::gpu::MatrixLayout::Order::kColumnMajor)
+         ? se::blas::Transpose::kNoTranspose : se::blas::Transpose::kTranspose;
+    auto vtype = se::gpu::AsBlasDataType(m.dtype);
+    // unsupported datatypes will be handled by the underlying BLAS layer
+    // which performs one more conversion anyway from gpu::DataType to native
+    // cu/roc BLAS datatype
+    auto type = vtype.ok() ? vtype.value() : 
+       static_cast< se::blas::DataType >(std::numeric_limits< int32_t >::max());
+
+    return se::gpu::MatrixDescriptor{
+      data, m.leading_dim_stride, m.batch_stride, type, trans
+    };
+  };
+  // make a local copy to prevent modification of layouts, 
+  // but maybe we can modify them once instead during creation ?
+  auto lhs = lhs_layout, rhs = rhs_layout, out = output_layout;
 
-// This struct contains the metadata of a matrix, e.g., its base address and
-// dimensions.
-struct MatrixDescriptor {
-  se::DeviceMemoryBase data;
-  int64_t leading_dim_stride;
-  int64_t batch_stride;
-  se::blas::Transpose transpose;
-
-  template <typename T>
-  se::DeviceMemory<T> cast() const {
-    return se::DeviceMemory<T>(data);
+  bool must_swap_operands = MakeOutputColumnMajor(lhs, rhs, out);
+  if (must_swap_operands) {
+    std::swap(lhs_buf, rhs_buf);
   }
-};
 
-se::blas::Transpose AsBlasTranspose(MatrixLayout::Order order) {
-  // BLAS is column-major by default.
-  return (order == MatrixLayout::Order::kColumnMajor)
-             ? se::blas::Transpose::kNoTranspose
-             : se::blas::Transpose::kTranspose;
+   auto comp_vtype = se::gpu::GetBlasComputationType(lhs.dtype, out.dtype,
+            se::blas::kDefaultComputePrecision);
+  // unsupported datatypes will be handled by the underlying BLAS layer
+  auto comp_type = comp_vtype.ok() ? comp_vtype.value() :
+        static_cast< se::blas::ComputationType >(
```

Comment:
I think we shouldn't do this. Can we instead return an error directly here?
(I know this will mean that create_matrix_descriptor will be a bit more difficult to call, because we will have to check for errors.)